tests/director/tests/application_templates_api_test.go

```diff
@@ -584,6 +584,74 @@ func TestDeleteApplicationTemplate(t *testing.T) {
 	saveExample(t, deleteApplicationTemplateRequest.Query(), "delete application template")
 }
 
+func TestDeleteApplicationTemplateBeforeDeletingAssociatedApplicationsWithIt(t *testing.T) {
+	//GIVEN
+	ctx := context.TODO()
+	nameJSONPath := "$.name-json-path"
+	displayNameJSONPath := "$.display-name-json-path"
+	appTemplateName := createAppTemplateName("template")
+	appTemplateInput := fixAppTemplateInputWithDefaultDistinguishLabel(appTemplateName)
+	appTemplateInput.ApplicationInput.Description = ptr.String("test {{display-name}}")
+	appTemplateInput.Placeholders = []*graphql.PlaceholderDefinitionInput{
+		{
+			Name:        "name",
+			Description: ptr.String("name"),
+			JSONPath:    &nameJSONPath,
+		},
+		{
+			Name:        "display-name",
+			Description: ptr.String("display-name"),
+			JSONPath:    &displayNameJSONPath,
+		},
+	}
+
+	tenantId := tenant.TestTenants.GetDefaultTenantID()
+
+	t.Log("Creating application template")
+	appTemplate, err := fixtures.CreateApplicationTemplateFromInput(t, ctx, certSecuredGraphQLClient, tenantId, appTemplateInput)
+	defer fixtures.CleanupApplicationTemplate(t, ctx, certSecuredGraphQLClient, tenantId, appTemplate)
+	require.NoError(t, err)
+	require.Equal(t, conf.SubscriptionConfig.SelfRegRegion, appTemplate.Labels[tenantfetcher.RegionKey])
+
+	appFromTemplate := graphql.ApplicationFromTemplateInput{TemplateName: appTemplateName, Values: []*graphql.TemplateValueInput{
+		{
+			Placeholder: "name",
+			Value:       "new-name",
+		},
+		{
+			Placeholder: "display-name",
+			Value:       "new-display-name",
+		}}}
+	appFromTemplateGQL, err := testctx.Tc.Graphqlizer.ApplicationFromTemplateInputToGQL(appFromTemplate)
+	require.NoError(t, err)
+	createAppFromTemplateRequest := fixtures.FixRegisterApplicationFromTemplate(appFromTemplateGQL)
+	outputApp := graphql.ApplicationExt{}
+
+	t.Logf("Creating application from application template with id %s", appTemplate.ID)
+	err = testctx.Tc.RunOperation(ctx, certSecuredGraphQLClient, createAppFromTemplateRequest, &outputApp)
```

Comment:
I think we should have a defer here to unregister the app. Because if something fails between line 631 and line 649, then the app will stay in the db as garbage, as well as the app template because the differ [here](https://github.com/kyma-incubator/compass/pull/2980/files#diff-cd91fb82642cfc9f5b55639748e76cc678d32a6a96513051cc75efae61540ce7R612) will fail due to the FK restriction.

Reply:
We might think about adapting other tests that create app from template like "TestRegisterApplicationFromTemplate" for example, in order to guarantee that other tests will not produce garbage.